x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,404 @@
+	if config.AccessKeyID != "" && config.SecretAccessKey != "" {
+		awsConfig := defaults.Config()
+		awsCredentials := awssdk.Credentials{
+			AccessKeyID:     config.AccessKeyID,
+			SecretAccessKey: config.SecretAccessKey,
+		}
+		if config.SessionToken != "" {
+			awsCredentials.SessionToken = config.SessionToken
+		}
+
+		awsConfig.Credentials = awssdk.StaticCredentialsProvider{
+			Value: awsCredentials,
+		}
+		return awsConfig, nil
+	}
+
+	// If accessKeyID and secretAccessKey is not given, then load from default config
+	// Please see https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html
+	// with more details.
+	if config.ProfileName != "" {
+		return external.LoadDefaultAWSConfig(
+			external.WithSharedConfigProfile(config.ProfileName),
+		)
+	}
+	return external.LoadDefaultAWSConfig()
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.logger.Debugf("s3", "Run s3 input with queueURLs: %+v", p.config.QueueURLs)
+	for _, queueURL := range p.config.QueueURLs {
+		regionName, err := getRegionFromQueueURL(queueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %s", queueURL)
+			continue
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		// update message visibility timeout if it's taking longer than 1/2 of
+		// visibilityTimeout to make sure filebeat can finish reading
+		changeMessageVisibilityInput := &sqs.ChangeMessageVisibilityInput{
+			QueueUrl:          &queueURL,
+			VisibilityTimeout: &visibilityTimeout,
+		}
+
+		// receive messages
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &queueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+		output, err := req.Send()
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			continue
+		}
+
+		// process messages received from sqs
+		if len(output.Messages) > 0 {
+			var wg sync.WaitGroup
+			numMessages := len(output.Messages)
+			wg.Add(numMessages)
+
+			for i := range output.Messages {
+				done := make(chan struct{})
+				// launch goroutine to handle each message from sqs
+				go func(message sqs.Message) {
+					defer wg.Done()
+					defer close(done)
+
+					s3Infos, err := handleMessage(message, p.config.BucketNames)
+					if err != nil {
+						p.logger.Error(err.Error())
+					}
+
+					// read from s3 object and create event for each log line
+					p.readS3Object(svcS3, s3Infos)
+
+					// delete message after events are sent
+					err = deleteMessage(queueURL, *message.ReceiptHandle, svcSQS)
+					if err != nil {
+						p.logger.Error(errors.Wrap(err, "deleteMessages failed"))
+					}
+				}(output.Messages[i])
+
+				go func(message sqs.Message) {
+					for {
+						select {
+						case <-done:
```

Comment:
we should return here, to escape the `for` loop and finish the goroutine

Reply:
Yes you are right, this is not a part of the previous function anymore. I added simply a `return` here in the new commit. Do I need to create more wait groups and add `defer wg.Done()` in this `go func` just to be more sure? I tested with `runtime.NumGoroutine` to show how many current goroutines exist and it shows with simply a `return`, the new goroutines get closed.

Reply:
I don't think the main loop needs to wait for these to be closed, as they are only handling the keep alive of messages. These goroutines are also guaranteed to close, as `done` channel is closed in all cases. So this LGTM as it is :+1:

Reply:
👍 Thanks for all the help here!!